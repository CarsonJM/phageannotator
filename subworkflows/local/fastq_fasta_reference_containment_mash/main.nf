//
// De novo identification of viral sequences in assemblies
//

include { MASH_SKETCH as MASH_SKETCH_ASSEMBLIES } from '../../../modules/nf-core/mash/sketch/main'
include { MASH_SKETCH as MASH_SKETCH_REFERENCES } from '../../../modules/nf-core/mash/sketch/main'
include { MASH_PASTE                            } from '../../../modules/local/mash/paste/main'
include { MASH_SCREEN                           } from '../../../modules/nf-core/mash/screen/main'
include { APPEND_SCREEN_HITS                    } from '../../../modules/local/append_screen_hits/main'
include { CAT_CAT as CAT_MASH_SCREEN            } from '../../../modules/nf-core/cat/cat/main'

workflow FASTQ_FASTA_REFERENCE_CONTAINMENT_MASH {
    take:
    fastq_gz    // [ [ meta] , [ 1.fastq.gz, 2.fastq.gz ]   , input reads (mandatory)
    fasta_gz    // [ [ meta] , fasta.gz ]                   , input assemblies (mandatory)

    main:
    ch_versions = Channel.empty()

    //
    // MODULE: Create a sketch of assemblies
    //
    ch_assembly_sketch_msh = MASH_SKETCH_ASSEMBLIES ( fasta_gz ).mash
    ch_versions = ch_versions.mix(MASH_SKETCH_ASSEMBLIES.out.versions.first())

    // if reference based identification requested, a reference FASTA file must be included
    if ( !params.skip_reference_containment && !params.reference_virus_fasta ) {
        error "[nf-core/phageannotator] ERROR: reference containment requested, but no --reference_virus_fasta provided"
    }

    // if provided, use reference sketch. If not, create one
    if ( params.reference_virus_sketch ) {
        ch_reference_sketch_msh = [ [ id: 'reference' ], file(params.reference_virus_sketch, checkIfExists: true) ]
    } else {
        //
        // MODULE: Create sketch of reference viruses
        //
        ch_reference_sketch_msh = MASH_SKETCH_REFERENCES ( [ [ id: 'reference' ], file(params.reference_virus_fasta, checkIfExists: true) ] ).mash
        ch_versions = ch_versions.mix(MASH_SKETCH_REFERENCES.out.versions.first())
    }

    //
    // MODULE: Combine assembly and reference sketches
    //
    ch_combined_sketch_msh = MASH_PASTE ( ch_assembly_sketch_msh, ch_reference_sketch_msh ).msh
    ch_versions = ch_versions.mix(MASH_PASTE.out.versions.first())

    // join reads and combined sketch within samples
    ch_mash_screen_input = fastq_gz.join(ch_combined_sketch_msh, by:0 )
    //
    // MODULE: Identify contained genomes
    //
    ch_mash_screen_tsv = MASH_SCREEN ( ch_mash_screen_input ).screen
    ch_versions = ch_versions.mix(MASH_SCREEN.out.versions.first())

    // join mash screen and assembly fasta within samples
    ch_append_screen_hits_input = ch_mash_screen_tsv.join( fasta_gz, by:0 )
    // create channel for reference virus fasta
    ch_reference_virus_fasta_gz = [ [ id: 'reference' ], file(params.reference_virus_fasta, checkIfExists: true) ]
    //
    // MODULE: Append screen hits to assemblies
    //
    ch_assembly_w_references_fasta_gz = APPEND_SCREEN_HITS ( ch_append_screen_hits_input, ch_reference_virus_fasta_gz ).assembly_w_screen_hits
    ch_versions = ch_versions.mix(APPEND_SCREEN_HITS.out.versions.first())

    //
    // MODULE: Combine mash screen outputs across samples
    //
    ch_combined_mash_screen_tsv = CAT_MASH_SCREEN( ch_mash_screen_tsv.map{ [ [ id:'all_samples' ], it[1] ] }.groupTuple() ).file_out
    ch_versions = ch_versions.mix(CAT_MASH_SCREEN.out.versions.first())

    emit:
    assembly_w_references_fasta_gz  = ch_assembly_w_references_fasta_gz // [ [ meta] , fasta.gz ] , concatenated assemblies and contained references
    versions                        = ch_versions                       // [ versions.yml ]
}
