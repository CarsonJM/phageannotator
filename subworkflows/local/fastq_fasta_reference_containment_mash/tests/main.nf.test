nextflow_workflow {

    name "Test subworkflow: FASTQ_FASTA_REFERENCE_CONTAINMENT_MASH"
    script "subworkflows/local/fastq_fasta_reference_containment_mash/main.nf"
    workflow "FASTQ_FASTA_REFERENCE_CONTAINMENT_MASH"
    tag "subworkflows"
    tag "subworkflows_local"
    tag "fastq_fasta_reference_containment_mash"
    tag "fastq_fasta_reference_containment_mash_default"


    test("Parameters: default") {

        when {
            params {
                outdir                  = "$outputDir"
            }
            workflow {
                """
                input[0] = Channel.of(
                    [ [ id:'test' ], [ file(params.test_data['bacteroides_fragilis']['illumina']['test1_1_fastq_gz'], checkIfExists:true), file(params.test_data['bacteroides_fragilis']['illumina']['test1_2_fastq_gz'], checkIfExists: true) ] ],
                    [ [ id:'test2' ], [ file(params.test_data['bacteroides_fragilis']['illumina']['test2_1_fastq_gz'], checkIfExists:true), file(params.test_data['bacteroides_fragilis']['illumina']['test2_2_fastq_gz'], checkIfExists: true) ] ]
                )
                input[1] = Channel.of([ [ id:'test' ], [ file(params.test_data['bacteroides_fragilis']['genome']['genome_fna_gz'], checkIfExists:true) ] ], [ [ id:'test2' ], [ file(params.test_data['bacteroides_fragilis']['genome']['genome_fna_gz'], checkIfExists:true) ] ])
                input[2] = Channel.of([ [ id:'reference_viruses' ], [ file(params.test_data['bacteroides_fragilis']['illumina']['test1_contigs_fa_gz'], checkIfExists:true) ] ])
                input[3] = null
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() }
            )
        }
    }
}
