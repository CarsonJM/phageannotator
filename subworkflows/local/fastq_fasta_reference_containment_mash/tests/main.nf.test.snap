{
    "full_fastq_fasta_reference_containment_mash": {
        "content": [
            [
                [
                    [
                        "reference_viruses.msh:md5,1603390852e55b99abee283a2490f070"
                    ]
                ]
            ]
        ],
        "timestamp": "2023-09-14T21:32:24+0000"
    }
}